reactive/kubernetes_master.py

```diff
@@ -851,30 +852,8 @@ def push_service_data():
     public interface '''
     kube_api = endpoint_from_flag('kube-api-endpoint.available')
 
-    # Note that we do not need to worry about the loadbalancer case because
-    # the worker charm will be related to the loadbalancer in that case and
-    # get the information from there instead of the kubernetes-master charm.
-
-    # if the user gave us IPs for the load balancer, assume they know
-    # what they are talking about and use that instead of our information.
-    address = None
-    forced_lb_ips = hookenv.config('loadbalancer-ips').split()
-    hacluster = endpoint_from_flag('ha.connected')
-    if forced_lb_ips:
-        address = forced_lb_ips
-    elif hacluster:
-        vips = hookenv.config('ha-cluster-vip').split()
-        dns_record = hookenv.config('ha-cluster-dns')
-        if vips:
-            # each worker unit will pick one based on unit number
-            address = vips
-        elif dns_record:
-            address = dns_record
-
-    if address:
-        kube_api.configure(6443, address, address)
-    else:
-        kube_api.configure(6443)
+    address, port = kubernetes_master.get_api_endpoint(kube_api.relations[0])
+    kube_api.configure(port, address, address)
```

Comment:
Is this the same behavior as before? We only set address if it was forced or we had hacluster before. get_api_endpoint now sets it to hookenv.unit_public_address as a default. configure defaults to hookenv.ingress_address. Will this break spaces?

Reply:
@hyperbolic2346 beat me to it, but here it is anyway.

Grabbing `kube_api.relations[0]` here seems wrong. Before, in the non-`forced_lb_ips`, non-`hacluster` case, we would invoke `kube_api.configure(6443)` with no address and let the http interface fall back to the ingress address for each relation: https://github.com/juju-solutions/interface-http/blob/632131b1f122daf6fb601fd4c9f1e4dbb1a92e09/provides.py#L53-L54.

With this change, in the case described above, every relation will be configured with the ingress address of the -first- relation.

Isn't it possible for each relation to have a different ingress address? I would expect this to break in scenarios involving multiple network spaces or cross-model relations.

Reply:
Even worse, it looks like when the loadbalancer relation is connected, kubernetes-master will send the loadbalancer's address over the kube-api-endpoint relation instead of its own. The kubeapi-load-balancer charm uses info from that relation to determine where to proxy traffic to the masters, but the masters will be telling it to proxy traffic to itself.

Reply:
Also, in the forced_lb_ips and hacluster cases, this would previously send a list of addresses to the http relation. Now it only sends one.

Reply:
I was assuming a single LB relation, but you're right that we shouldn't make that assumption if we can avoid it.  As for sending the LB address, I read the comment that was removed to mean that the value would be ignored if the LB was connected, but I see that I probably understood that wrong.  I think I'll have to exclude this one from the refactor as well, meaning it will only be touching I think one other place.  Frustrating that this logic is slightly different in each place making it difficult to factor up.